docs/troubleshooting/vm.md

```diff
@@ -243,3 +243,174 @@ After the pod is successfully deleted, the `Start` button becomes visible again
+```
+
+This is a CSI related file is an empty file instead of the expected device file. Then the containerd denied the `CreateContainer` request.
+
+```
+$ ls /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/ -alth
+total 8.0K
+drwxr-x--- 2 root root 4.0K Feb 22 15:10 .
+-rw-r--r-- 1 root root    0 Feb 22 14:28 aa851da3-cee1-45be-a585-26ae766c16ca
+-rw-r--r-- 1 root root    0 Feb 22 14:07 20291c6b-62c3-4456-be8a-fbeac118ec19
+drwxr-x--- 4 root root 4.0K Feb 22 14:06 ..
+-rw-r--r-- 1 root root    0 Feb 21 15:48 4333c9fd-c2c8-4da2-9b5a-1a310f80d9fd
+-rw-r--r-- 1 root root    0 Feb 21 09:18 becc0687-b6f5-433e-bfb7-756b00deb61b
+
+$file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/20291c6b-62c3-4456-be8a-fbeac118ec19
+: empty
+```
+
+By contrast, another running VM's related file is the expected device file.
+
+```
+$ ls  /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-732f8496-103b-4a08-83af-8325e1c314b7/ -alth
+total 8.0K
+drwxr-x--- 2 root root  4.0K Feb 21 10:53 .
+drwxr-x--- 4 root root  4.0K Feb 21 10:53 ..
+brw-rw---- 1 root root 8, 16 Feb 21 10:53 4883af80-c202-4529-a2c6-4e7f15fe5a9b
+```
+
+#### Root Cause
+
+The kubelet calls `NodePublishVolume` for the new pod after node/cluster rebooting without first calling `NodeStageVolume`, and the Longhorn CSI plugin bind mounts the regular file at staging target path (previously used by the deleted pod) to target path and returns success.
+
+### Workaround
+
+Cluster level operation:
+
+1. Find the backing pods of the affected VMs and the related Longhorn volumes.
+
+    ```
+    $ kubectl get pods
+    NAME                      READY   STATUS                 RESTARTS   AGE
+    virt-launcher-vm1-nxfm4   0/2     CreateContainerError   0          7m11s
+
+    $ kubectl get pvc -A
+    NAMESPACE                  NAME                       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS           AGE
+    default                    vm1-disk-0-9gc6h           Bound    pvc-f1798969-5b72-4d76-9f0e-64854af7b59c   1Gi        RWX            longhorn-image-fxsqr   7d22h
+    ```
+
+1. [Stop](#vm-general-operations) the affected VMs from Harvester UI.
+
+   The VM may stuck in `Stopping`, continue the next step.
+
+1. Delete the backing pods forcely.
+
+    ```
+    $ kubectl delete pod virt-launcher-vm1-nxfm4 --force
+    Warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
+    pod "virt-launcher-vm1-nxfm4" force deleted
+    ```
+
+    The VM is off now.
+
+    ![](/img/v1.3/troubleshooting/vm-is-off.png)
+
+Node level operation, node by node:
+
+1. [Cordon](../host/host.md#cordoning-a-node) a node.
```

Comment:
I cordon the node in my workaround because the Deployment I am using will immediately reschedule a pod to the node if I delete one. This prevents the CSI flow from ever getting unstuck.

If you forcefully delete the pods above, will they be potentially rescheduled to the node? If so, should we cordon first?

If there is no danger that the pods will be rescheduled to the node, is cordoning necessary?

Reply:
The idea is to stop & clean the affected VM & POD at first, avoid they are scheduled/redeployed to other nodes accidently. 

Then cordon the node to unmount the volumes.

The `cordon` maybe optional if all affected VMs have been stopped&cleaned.

Reply:
I need to further consider the scenario or Rancher managed downstream clusters, if a VM is stopped from Harvester UI, will this VM be re-enabled or recreated automatically? That may lead to kind of race.

Reply:
Ok. I certainly have no concerns leaving the cordon in if we are unsure. I think I mostly wanted to be sure that the VM couldn't be rescheduled to the node BEFORE the cordon. If Harvester will not do that, all good.